packages/nx/src/plugins/js/project-graph/build-dependencies/target-project-locator.ts

```diff
@@ -556,14 +569,29 @@ export class TargetProjectLocator {
 
       while (dir !== dirname(dir)) {
         const packageJsonPath = join(dir, 'package.json');
-        try {
-          const parsedPackageJson = readJsonFile(packageJsonPath);
-          // Ensure the package.json contains the "name" and "version" fields
-          if (parsedPackageJson.name && parsedPackageJson.version) {
-            return parsedPackageJson;
+        if (this.packageJsonResolutionCache.has(packageJsonPath)) {
+          return this.packageJsonResolutionCache.get(packageJsonPath);
+        }
+        if (existsSync(packageJsonPath)) {
+          try {
+            const parsedPackageJson = readJsonFile(packageJsonPath);
+            // Ensure the package.json contains the "name" and "version" fields
+            if (parsedPackageJson.name && parsedPackageJson.version) {
+              this.packageJsonResolutionCache.set(
+                packageJsonPath,
+                parsedPackageJson
+              );
+              return parsedPackageJson;
+            } else {
+              this.packageJsonResolutionCache.set(packageJsonPath, null);
+              return null;
+            }
+          } catch {
+            // Package.json is invalid, keep traversing
```

Comment:
```
this.packageJsonResolutionCache.set(packageJsonPath, null);
```

Save null here so that we don't have to retry parsing jsons that we cannot parse or ones that don't exist